synapse/lib/fifo.py

```diff
@@ -0,0 +1,389 @@
+import os
+import struct
+import logging
+import threading
+import collections
+
+import synapse.glob as s_glob
+import synapse.common as s_common
+import synapse.eventbus as s_eventbus
+
+import synapse.lib.atomic as s_atomic
+import synapse.lib.config as s_config
+import synapse.lib.msgpack as s_msgpack
+import synapse.lib.atomfile as s_atomfile
+
+logger = logging.getLogger(__file__)
+
+class Fifo(s_config.Config):
+
+    def __init__(self, conf, xmit=None):
+
+        s_config.Config.__init__(self)
+        self.setConfOpts(conf)
+        self.reqConfOpts()
+
+        self.maxsize = self.getConfOpt('fifo:file:maxsize')
+
+        self.lock = threading.Lock()
+        self.winds = s_eventbus.BusRef()
+
+        self.atoms = s_eventbus.BusRef(ctor=self._brefAtomCtor)
+        self.onfini(self.atoms.fini)
+
+        self.seqs = self._getFifoSeqs()
+        if not self.seqs:
+            self.seqs.append(0)
+
+        # open our append atom
+        lseq = self.seqs[-1]
+        self.atom = self.atoms.gen(lseq)
+
+        # our next expected sequence num
+        self.nseq = lseq + self.atom.size
+
+        # open our "default" state file
+        path = self._getPathJoin('default.state')
+        self.wind = Window(self, path, xmit=xmit)
+        self.onfini(self.wind.fini)
+
+        task = s_glob.sched.loop(2, self.flush)
+        self.onfini(task.fini)
+
+        self._cullAtomBase(self.wind.bseq)
+
+    def ack(self, seqn):
+        '''
+        Acknowledge (and cull) an item in the sequence.
+        '''
+        retn = self.wind.ack(seqn)
+
+        if retn:
+            self._cullAtomBase(self.wind.bseq)
+
+        return retn
+
+    def _cullAtomBase(self, nseq):
+
+        with self.lock:
+
+            while self.seqs[0] < nseq:
+                fseq = self.seqs.pop(0)
+                path = self._getSeqPath(fseq)
+                os.unlink(path)
+
+    def resync(self, xmit=None):
+        '''
+        Re-synchronize this Fifo by sending all window entries.
+        (optionally specify a new xmit callback)
+
+        Args:
+            xmit (func): The fifo xmit() callback.
+
+        Example:
+
+            def xmit(qent):
+                seqn, nseq, item = qent
+                dostuff()
+
+            fifo.resync(xmit=xmit)
+        '''
+        return self.wind.resync(xmit=xmit)
```

Comment:
The xmit use here feels unclear.  We can ``__init__`` a fifo with a xmit and then the resync call won't use that xmit.  However, this will overwrite the self.wind._xmit which was previously set during construction.

Reply:
This is def on purpose.  a resync() call would occur, for example, on a socket close and reconnect.